lib/Dialect/TritonNvidiaGPU/Transforms/FenceInsertion.cpp

```diff
@@ -50,21 +55,62 @@ struct FenceInsertionPass
                                .getEncoding()
                                .dyn_cast<ttg::MmaEncodingAttr>();
         auto isHopperEncoding = mmaEncoding && mmaEncoding.isHopper();
-        if (isHopperEncoding && (a.getDefiningOp<ttg::ConvertLayoutOp>() &&
-                                 ttg::isSharedEncoding(a)) ||
-            (b.getDefiningOp<ttg::ConvertLayoutOp>() &&
-             ttg::isSharedEncoding(b))) {
-
-          // TODO: check whether cluster fence is needed
-          OpBuilder builder(op);
+        if (isHopperEncoding &&
+            (dependOnSharedEncOperand(a) || dependOnSharedEncOperand(b))) {
           builder.create<ttng::FenceAsyncSharedOp>(op->getLoc(),
                                                    false /*bCluster*/);
         }
       }
     });
   }
-};
 
+private:
+  bool dependOnSharedEncOperand(Value operand) {
+    static DenseSet<std::pair<Operation *, unsigned>> trace;
+    auto op = operand.getDefiningOp();
+    // reach convertlayout
+    if (op && isa<ttg::ConvertLayoutOp>(op) && ttg::isSharedEncoding(operand))
+      return true;
+    // root and not BlockArgument
+    if (!op && !isa<BlockArgument>(operand))
+      return false;
```

Comment:
```
if (isa<DotOp/DotAsyncOp>(op) and op.getEncoding().isHopper)
  return false
```

Do we need to add something like this? to prevent emit redundant fences?